src/Services/Settings.js

```diff
@@ -85,11 +83,50 @@ class Settings {
     // Ensure new settings variables get added by merging all the settings,
     // where the current values take precedence. 
     let currentSettings = this._getAllRaw()
+
+    // Add any non-additive settings changes here by creating a function which
+    // handles the settings change in question.
+    currentSettings = this.migrateMnemonicSettings(currentSettings);
+
     currentSettings = _.merge({}, initialSettings, currentSettings);
 
     // Apply the merged settings
     this.setAll(currentSettings);
   }
+
+  migrateMnemonicSettings(currentSettings) {
+    // If we're migrating a settings file from before we used a persistent,
+    // randomly generated mnemonic by default, randomizeMnemonic on start will
+    // be undefined.
+    if (currentSettings.randomizeMnemonicOnStart === undefined) {
+
+      // Before we added the randomizeMnemonicOnStart flag, the absence of a
+      // mnemonic meant that we wanted a random one one each run. We want to
+      // preserve this preference.
+      if (currentSettings.server.mnemonic === "") {
+        currentSettings.randomizeMnemonicOnStart = true;
+      } else if (currentSettings.server.mnemonic === oldDefaultMnemonic || !currentSettings.server.mnemonic) {
+
+        // This will cause a new mnemonic to be generated and persisted only in
+        // the case when the old default mnemonic was being used.
+        currentSettings.server.mnemonic = null;
+      }
+    }
+
+    return currentSettings;
+  }
+
+  /**
+   * Called when a new mnemonic is read back from the underlying chain, will
+   * persist this new mnemonic if it differs from the one stored and if
+   * randomizeMnemonicOnStart is false.
+   */
+  handleNewMnemonic(mnemonic) {
```

Comment:
Like before, this is a setter that doesn't always set things. Shouldn't the logic (if statement) be moved to the caller? https://github.com/trufflesuite/ganache/pull/378/files#diff-1c90ff38a08209f9ebd4d05d1e43358eR113